app/src/main/java/com/eveningoutpost/dexdrip/Models/DateUtil.java

```diff
@@ -1,81 +1,44 @@
+ *
  * Created by mike on 30.12.2015.
- */
-
-/**
- * The Class DateUtil. A simple wrapper around SimpleDateFormat to ease the handling of iso date string &lt;-&gt; date obj
- * with TZ
+ *
  */
 public class DateUtil {
 
     private static final String FORMAT_DATE_ISO = "yyyy-MM-dd'T'HH:mm:ss'Z'"; // eg 2017-03-24T22:03:27Z
-    private static final String FORMAT_DATE_ISO2 = "yyyy-MM-dd'T'HH:mm:ssZ"; // eg 2017-03-27T17:38:14+0300
-    private static final String FORMAT_DATE_ISO3 = "yyyy-MM-dd'T'HH:mmZ"; // eg 2017-05-12T08:16-0400
+    private static final Pattern pDateFix = Pattern.compile("^(\\d{4}-\\d{2}-\\d{2})[ T](\\d{2}:\\d{2}:\\d{2})(?:.\\d+Z)*[^Z]?$");
+    private static final Pattern pOffsetFix = Pattern.compile("([+-]\\d{2})(\\d{2})$");
 
     /**
      * Takes in an ISO date string of the following format:
-     * yyyy-mm-ddThh:mm:ss.ms+HoMo
+     * yyyy-MM-dd'T'HH:mm:ss.SSS(Z|+/-HH:mm)
+     *
+     * The milliseconds and time zone/offset values are optional.
      *
-     * @param isoDateString the iso date string
-     * @return the date
-     * @throws Exception the exception
+     * @param isoDateString the iso date string to parse
+     * @return the date object
      */
-    private static Date fromISODateString(String isoDateString)
-            throws Exception {
-        SimpleDateFormat f = new SimpleDateFormat(FORMAT_DATE_ISO);
-        f.setTimeZone(TimeZone.getTimeZone("UTC"));
-        return f.parse(isoDateString);
-    }
-
-    
-    private static Date fromISODateString4(String isoDateString)
-            throws Exception {
-        DateTime dt = ISODateTimeFormat.dateTime().parseDateTime(isoDateString);
-        return dt.toDate();
-    }
-    
-    private static Date fromISODateString3(String isoDateString)
-            throws Exception {
-        try {
-            SimpleDateFormat f = new SimpleDateFormat(FORMAT_DATE_ISO3);
-            f.setTimeZone(TimeZone.getTimeZone("UTC"));
-            return f.parse(isoDateString);
-        } catch (java.text.ParseException e) {
-            return fromISODateString4(isoDateString);
-        }
-        
-    }
-
-    private static Date fromISODateString2(String isoDateString)
-            throws Exception {
-        try {
-            SimpleDateFormat f = new SimpleDateFormat(FORMAT_DATE_ISO2);
-            f.setTimeZone(TimeZone.getTimeZone("UTC"));
-            return f.parse(isoDateString);
-        } catch (java.text.ParseException e) {
-            return fromISODateString3(isoDateString);
-        }
-    }
-
-    public static Date tolerantFromISODateString(String isoDateString)
-            throws Exception {
-        try {
-            return fromISODateString(isoDateString.replaceFirst("\\.[0-9][0-9][0-9]Z$", "Z"));
-        } catch (java.text.ParseException e) {
-            return fromISODateString2(isoDateString);
+    public static Date tolerantFromISODateString(String isoDateString) {
```

Comment:
I haven't examined this in any detail but I'm kind of nervous about modifying this code. Are you confident the new unit tests cover whatever might have been seen and be accepted by this method?

Reply:
I've added an additional test case to cover date/time strings without seconds but offset value (FORMAT_DATE_ISO3). Now all former string formats are covered and the regex fixes much more "wrong" date/time strings. See https://regex101.com/r/utF21U/1 for a clean impression.